Fixes #2944, #2947 - Read experiment variation values from envvars

miketaylr opened the pull request:
This feels maybe slightly hacky -- we can't safely store a tuple in an environment variable, so it build one up from something like `0 20`. Open to other ideas.

And add tests.

r? @johngian @karlcow

miketaylr commented:
Oops, wrong bug number. :) Will fix the commits.

karlcow commented on tests/unit/test_config.py:
```diff
@@ -53,6 +54,19 @@ def test_update_status_config(self):
+        self.assertRaises(Exception, get_variation_from_env, '1 ')
+        self.assertTupleEqual((1, 2), get_variation_from_env('1 2'))
+        self.assertTupleEqual((1, 2), get_variation_from_env(' 1 2 '))
+
```
❤️ I tested '0010    02' 
And this is working too.

karlcow commented on tests/unit/test_config.py:
```diff
@@ -53,6 +54,19 @@ def test_update_status_config(self):
+    def test_get_variation_from_env(self):
+        """Make sure it doesn't blow up when it shouldn't, and does when
+        it should.
+        """
```
nit: Probably a different docstrings ;) I know my docstrings are boring. :p 

```suggestion
"""
Check the conversion from string to tuples of two int.
"""
```

karlcow commented on docs/ab-testing.md:
````diff
@@ -1,10 +1,23 @@
+
+```
+export FORM_V1_VARIATION = "0 20"
+export FORM_V2_VARIATION = "20 100"
````
Maybe not related to this PR, but maybe @johngian can explain. Why do we need to define the two alternatives when one seems to be the differential of the other one.

johngian replied:
Just easier to implement, more explicit. We can change it in the future if want more A/B experiments and need a more flexible config.

johngian commented:
I was thinking of a different approach that
a) Would work for other experiments as well
b) Has less hardcoded values

can be something like that: https://github.com/johngian/webcompat.com/commit/0662bf14e81fd1b8b4a08e484fc97663e420a80f

The reasons that I came up with this is that if something goes very bad and we need to override the A/B testing cookies (in our case `exp`), this would require a code deployment. Where, if we manage the whole A/B config definition we can just change the value there.

johngian commented on config/__init__.py:
```diff
@@ -200,14 +219,20 @@ def update_status_config(milestones_content):
+# By default, it will serve v1 100% of the time.
+FORM_V1_VARIATION = os.environ.get('FORM_V1_VARIATION') or '0 100'
+FORM_V2_VARIATION = os.environ.get('FORM_V2_VARIATION') or '100 100'
+EXP_MAX_AGE = os.environ.get('EXP_MAX_AGE', None)
```
max-age is defined as an integer so if we want to read this from the env var we should cast it to int (env vars are strings)

miketaylr commented:
> @miketaylr I haven't really thought about that since its more a once-off thing (writing the config i mean) and if you are **really** careful and pass it through a JSON linter you are probably good to go. It feels like writing a script to generate a JSON for the config its too much for our case.

thanks -- i think let's just go with this approach and i'll file a follow-up to simplify things via JSON parsing (more simple implementation, slightly higher risk). i'll work on addressing feedback.

miketaylr commented on config/__init__.py:
```diff
@@ -119,6 +119,32 @@ def update_status_config(milestones_content):
+
+    Raise an exception if the passed in envvar doesn't look something
+    like: '0 100'
+    """Convert a string to a tuple of integers.
```
Oops, looks like GitHub's accept suggestion thingy got confused. 💣

karlcow replied:
yeah the suggestion works for one line only.

miketaylr commented on config/__init__.py:
```diff
@@ -200,14 +219,20 @@ def update_status_config(milestones_content):
         'variations': {
-            'form-v1': (0, 100),
-            'form-v2': (100, 100)
+            'form-v1': get_variation_from_env(FORM_V1_VARIATION),
```
Looking again, I don't think this will break because we provide some default values `FORM_V1_VARIATION = os.environ.get('FORM_V1_VARIATION') or '0 100'` -- this should serve v1 100% of the time. We just need to add some logging here if it raises.

karlcow replied:
It will break take this scenario:

1. environment variable defined as `export FORM_V1_VARIATION='0 a'`
2. `FORM_V1_VARIATION = os.environ.get('FORM_V1_VARIATION', '0 100')` is valid and `FORM_V1_VARIATION` will be `0 a`
3. The code goes to `AB_EXPERIMENTS` which contains `'form-v1': get_variation_from_env(FORM_V1_VARIATION)`
4. This will raise an exception. There is no default value at this stage.

The default value applies only if environment variable doesn't exist.

miketaylr replied:
Ah. Right.

karlcow commented:
> > @miketaylr I haven't really thought about that since its more a once-off thing (writing the config i mean) and if you are **really** careful and pass it through a JSON linter you are probably good to go. It feels like writing a script to generate a JSON for the config its too much for our case.
> 
> thanks -- i think let's just go with this approach and i'll file a follow-up to simplify things via JSON parsing (more simple implementation, slightly higher risk). i'll work on addressing feedback.

the issue with a JSON file is that we are back to square one. As something which is dependent of a deploy (If I got what was proposed)

miketaylr commented:
> the issue with a JSON file is that we are back to square one. As something which is dependent of a deploy (If I got what was proposed)

I think the suggestion is to use a JSON *string*, not a file. So you export a var that contains JSON, rather than read it from disk. But I guarantee I will miquote the JSON and everything will fail. 😄

miketaylr commented on config/__init__.py:
```diff
@@ -119,6 +119,25 @@ def update_status_config(milestones_content):
+        # We want to create a tuple of integers from a string containing
+        # integers. Anything else should throw.
+        rv = tuple(int(x) for x in envvar_value.strip().split(' '))
+        if (len(rv) is not 2):
```
(note, it should be `!=` because we want to raise if the length is not 2, but i assume that's just a typo)

karlcow replied:
yup

miketaylr commented:
I think I addressed all feedback, r? @karlcow 

Right now, if `get_variation_from_env` raises because something is wacky, it will print to `stdout`. So we'll have to make sure we don't see any weird messages when restarting the app after updating values.

miketaylr commented:
OK, not 100% sure I love this implementation, but it works and the tests pass. 

(It doesn't feel very DRY, but I'm not sure I care. It would have been nice to be able to read the `VARIATIONS` and `DEFAULTS` dicts from the config during testing, rather than have to pass them in to be testable, but I don't think it's possible to refer to the config object inside of `get_variation` because at the point it gets called, the config object hasn't even been created.)

r? @karlcow

miketaylr commented:
(before merging, we should probably squash this into a single commit -- i'm happy to do so)

miketaylr commented:
This will also fix #2947 because by default the variables won't be set, and it will only serve v1 of the form (which tests rely on).

karlcow commented:
ok squashed committed. Let's run CircleCi and wait a bit.

karlcow commented:
merged
